questionpy_sdk/webserver/question_ui.py

```diff
@@ -0,0 +1,622 @@
+            index_str = int_to_roman(index).upper()
+        else:
+            index_str = str(index)
+
+        # Replace the index element with the new index string
+        new_text_node = etree.Element("span")  # Using span to replace the custom element
+        new_text_node.text = index_str
+
+        if index_element.tail:
+            new_text_node.tail = index_element.tail
+
+        index_element.getparent().replace(index_element, new_text_node)
+
+
+class QuestionUIRenderer:
+    XHTML_NAMESPACE: str = "http://www.w3.org/1999/xhtml"
+    QPY_NAMESPACE: str = "http://questionpy.org/ns/question"
+    question: etree._Element
+    placeholders: dict[str, str]
+
+    def __init__(self, xml: str, placeholders: dict[str, str], seed: Optional[int] = None) -> None:
+        self.seed = seed
+        self.xml = xml
+        self.placeholders = placeholders
+        self.question = etree.fromstring(xml.encode())
+
+    def get_metadata(self) -> QuestionMetadata:
+        """Extracts metadata from the question UI.
+
+        Returns:
+            QuestionMetadata: question_metadata
+        """
+        question_metadata = QuestionMetadata()
+        namespaces: dict[str, str] = {'xhtml': self.XHTML_NAMESPACE, 'qpy': self.QPY_NAMESPACE}
+
+        # Extract correct responses
+        for element in self.question.findall(".//qpy:formulation//*[@qpy:correct-response]",
+                                             namespaces=namespaces):
+            name = element.get("name")
+            if not name:
+                continue
+
+            if element.tag.endswith("input") and element.get("type") == "radio":
+                value = element.get("value")
+            else:
+                value = element.get(f"{{{self.QPY_NAMESPACE}}}correct-response")
+
+            if not value:
+                continue
+
+            question_metadata.correct_response[name] = value
+
+        # Extract other metadata
+        for element_type in ['input', 'select', 'textarea', 'button']:
+            for element in self.question.findall(f".//qpy:formulation//{{{self.XHTML_NAMESPACE}}}{element_type}",
+                                                 namespaces=namespaces):
+                name = element.get("name")
+                if not name:
+                    continue
+
+                question_metadata.expected_data[name] = "Any"
+                if element.get("required") is not None:
+                    question_metadata.required_fields.append(name)
+
+        return question_metadata
+
+    def render_general_feedback(self, attempt: Optional[dict] = None,
+                                options: Optional[QuestionDisplayOptions] = None) -> Optional[str]:
+        """Renders the contents of the `qpy:general-feedback` element or returns `None` if there is none.
+
+        Args:
+            attempt: Optional[dict]
+            options: Optional[QuestionDisplayOptions]
+
+        Returns:
+            Optional[str]
+        """
```

Comment:
Argument-Typen würdest du im Docstring mit `attempt (Optional[dict]): Beschreibung` angeben. Das ist aber gar nicht nötig (oder hilfreich), wenn eh die Signatur getypehinted ist.

```suggestion
        """Renders the contents of the `qpy:general-feedback` element or returns `None` if there is none."""
```